core/unit_test/view/TestMDSpanConversion.hpp

```diff
@@ -0,0 +1,393 @@
+                                   Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using natural_mdspan_type = typename Kokkos::Impl::MDSpanViewTraits<
+        typename view_type::traits>::mdspan_type;
+    using mapping_type       = MDSpanLayoutMapping;
+    using mdspan_layout_type = typename MDSpanLayoutMapping::layout_type;
+    using extents_type       = typename mapping_type::extents_type;
+    using mdspan_type =
+        Kokkos::mdspan<value_type, extents_type, mdspan_layout_type>;
+
+    static_assert(std::is_constructible_v<natural_mdspan_type, mdspan_type>);
+    static_assert(std::is_convertible_v<mdspan_type, natural_mdspan_type> ==
+                  std::is_convertible_v<mdspan_type, view_type>);
+
+    // Manually create an mdspan from ref so we have a valid pointer to play
+    // with
+    const auto &exts = mapping.extents();
+    auto mds         = mdspan_type{ref.data(), mapping};
+
+    auto test_view = view_type(mds);
+
+    ASSERT_EQ(test_view.data(), ref.data());
+    ASSERT_EQ(test_view.data(), mds.data_handle());
+    ASSERT_EQ(test_view.layout(), ref.layout());
+    for (std::size_t r = 0; r < mdspan_type::rank(); ++r) {
+      ASSERT_EQ(test_view.extent(r), ref.extent(r));
+      ASSERT_EQ(test_view.extent(r), exts.extent(r));
+    }
+  }
+
+  template <class MDSpanLayoutMapping, class ViewType>
+  static void test_conversion_to_mdspan(
+      const MDSpanLayoutMapping &ref_layout_mapping, ViewType v) {
+    using view_type           = ViewType;
+    using natural_mdspan_type = typename Kokkos::Impl::MDSpanViewTraits<
+        typename view_type::traits>::mdspan_type;
+    // test conversion operator to natural mdspan
+    {
+      natural_mdspan_type cvt = v;
+      ASSERT_EQ(cvt.data_handle(), v.data());
+      ASSERT_EQ(cvt.mapping(), ref_layout_mapping);
+    }
+    // test to_mdspan() returning natural mdspan
+    {
+      auto cvt = v.to_mdspan();
+      static_assert(std::is_same_v<natural_mdspan_type, decltype(cvt)>);
+      ASSERT_EQ(cvt.data_handle(), v.data());
+      ASSERT_EQ(cvt.mapping(), ref_layout_mapping);
+    }
+    // test conversion operator to different mdspan type
+    {
+      using mdspan_type = Kokkos::mdspan<
+          const typename natural_mdspan_type::element_type,
+          Kokkos::dextents<typename natural_mdspan_type::index_type,
+                           natural_mdspan_type::rank()>,
+          typename natural_mdspan_type::layout_type,
+          typename natural_mdspan_type::accessor_type>;
+      mdspan_type cvt = v;
+      ASSERT_EQ(cvt.data_handle(), v.data());
+      ASSERT_EQ(cvt.mapping(), ref_layout_mapping);
+    }
+  }
+
+  template <class MDSpanLayoutMapping, class ViewType, class AccessorType>
+  static void test_conversion_to_mdspan(
+      const MDSpanLayoutMapping &ref_layout_mapping, ViewType v,
+      const AccessorType &a) {
+    auto cvt = v.to_mdspan(a);
+    ASSERT_EQ(cvt.data_handle(), v.data());
+    ASSERT_EQ(cvt.mapping(), ref_layout_mapping);
```

Comment:
Should we add some basic static assertions such as value_type and rank are the same?